packages/events/src/events.ts

```diff
@@ -0,0 +1,264 @@
+// Copyright IBM Corp. 2018. All Rights Reserved.
+// Node module: @loopback/listener
+// This file is licensed under the MIT License.
+// License text available at https://opensource.org/licenses/MIT
+
+// tslint:disable-next-line:no-any
+export type Event = any;
+
+/**
+ * Async event listener function
+ */
+export type ListenerFunction = (
+  /**
+   * Event type
+   */
+  eventType: string,
+  /**
+   * Event data
+   */
+  event: Event,
+) => Promise<void>;
+
+/**
+ * Async event listener Object
+ */
+export interface ListenerObject {
+  /**
+   * An optional name of the listener
+   */
+  name?: string;
+  /**
+   * Process an event
+   * @param eventType Event type
+   * @param event Event data
+   */
+  listen(eventType: string, event: Event): Promise<void>;
+}
+
+/**
+ * A listener object or function
+ */
+export type Listener = ListenerObject | ListenerFunction;
+
+/**
+ * Emit an event to the given listener
+ * @param listener Listener
+ * @param eventType Event type
+ * @param event Event data
+ */
+function emitEvent(listener: Listener, eventType: string, event: Event) {
+  if (typeof listener === 'function') {
+    return listener(eventType, event);
+  }
+  if (typeof listener.listen === 'function') {
+    return listener.listen(eventType, event);
+  }
+  throw new Error(`Invalid listener: ${listener}`);
+}
+
+/**
+ * An map of listeners keyed by event types
+ */
+export interface ListenerMap {
+  [eventType: string]: Listener[];
+}
+
+/**
+ * A subscription
+ */
+export interface Subscription {
+  /**
+   * Cancel the subscription
+   */
+  cancel(): void;
+}
+
+/**
+ * An interface to describe an observable object
+ */
+export interface AsyncEventEmitter {
+  /**
+   * Get a list of listeners for the given source object and event type
+   * @param source Source object
+   * @param eventType Event type
+   */
+  getListeners(eventType: string): Listener[];
+
+  /**
+   * Subscribe to an event type
+   * @param eventType Event type
+   * @param listener An async event listener
+   */
+  subscribe(eventType: string, listener: Listener): Subscription;
```

Comment:
I think we should support `subscribeOnce` too.

Reply:
Added `once`.